README.md

## Android
<p float="left">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/android1.png" width="250">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/android2.png" width="250">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/android3.png" width="250">
</p>

## IOS
<p float="left">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/ios1.png" width="250">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/ios2.png" width="250">
  <img src="https://raw.githubusercontent.com/izedomi/flutter-monnify/main/screenshots/ios3.png" width="250">
</p>


## Flutter Monnify Package
A Flutter package for making payments via Monnify Payment Gateway. Android and iOS supported.

## Getting Started
To use this package, add `flutter_monnify` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).


## How to use

``` dart
import 'package:flutter_monnify/flutter_paystack.dart';

TransactionResponse? response = await Monnify().checkout(
      context, 
      monnifyPayload(),
 );

```

No other configuration required&mdash;the plugin works out of the box.

Customize the appbar to suit your project's theme color using the `AppConfig` object.

## Example
``` dart
import 'package:flutter/material.dart';
import 'package:flutter_monnify/flutter_monnify.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Monnify',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const PayWithMonnify(),
    );
  }
}

class PayWithMonnify extends StatefulWidget {
  const PayWithMonnify({super.key});

  @override
  State<PayWithMonnify> createState() => _PayWithMonnifyState();
}

class _PayWithMonnifyState extends State<PayWithMonnify> {
  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    double height = MediaQuery.of(context).size.height;
    return Scaffold(
        body: SizedBox(
      width: width,
      height: height,
      child: Center(
        child: TextButton(
          child: const Text("Pay With Monnify SDK"),
          onPressed: () async {
             TransactionResponse? response = await Monnify().checkout(
                context, monnifyPayload(),
                appBar: AppBarConfig(
                    titleColor: Colors.white, backgroundColor: Colors.red),
                toast: ToastConfig(
                    color: Colors.black, backgroundColor: Colors.red));

            //call the backend to verify transaction status before providing value
          },
        ),
      ),
    ));
  }

  Map<String, dynamic> monnifyPayload() {
    return {
      "amount": 100,
      "currency": "NGN",
      "reference": DateTime.now().toIso8601String(),
      "customerFullName": "John Doe",
      "customerEmail": "customer@gmail.com",
      "apiKey": "MK_TEST_595UN92CCV",
      "contractCode": "8628159341",
      "paymentDescription": "Lahray World",
      "metadata": {"name": "Damilare", "age": 45},
      // "incomeSplitConfig": [
      //   {
      //     "subAccountCode": "MFY_SUB_342113621921",
      //     "feePercentage": 50,
      //     "splitAmount": 1900,
      //     "feeBearer": true
      //   },
      //   {
      //     "subAccountCode": "MFY_SUB_342113621922",
      //     "feePercentage": 50,
      //     "splitAmount": 2100,
      //     "feeBearer": true
      //   }
      // ],
      "paymentMethod": ["CARD", "ACCOUNT_TRANSFER", "USSD", "PHONE_NUMBER"],
    };
  }
}

```

The Monnify.checkout() returns a TransactionResponse with the following fields:

``` dart
  String? status;
  String? message;
  String? transactionReference;
  String? paymentReference;
  int? amount;
  String? currency;
  dynamic rawData;

```

The transaction is successful if `response.status` is `SUCCEESSFUL`. It is recommended to verify with the transaction status with the backend before providing balue

## Contributing, Issues and Bug Reports

The project is open to public contribution. Please feel very free to contribute.
Experienced an issue or want to report a bug? Please, [report it here](https://github.com/izedomi/flutter-monnify/issues). Remember to be as descriptive as possible.


## Need More Information?

For further info about Monnify's SDKs, including transaction status types, see the documentations for the [web](https://teamapt.atlassian.net/wiki/spaces/MON/pages/212008793/Monnify+Web+SDK) [Android](https://teamapt.atlassian.net/wiki/spaces/MON/pages/213909311/Monnify+Android+SDK) and [iOS](https://teamapt.atlassian.net/wiki/spaces/MON/pages/213909672/Monnify+iOS+SDK)






CHANGELOG.md
## 0.0.1

Initial Release.

## 0.0.2

Initial Release


lib/src/monnify_webview.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:webview_flutter_plus/webview_flutter_plus.dart';
import '../flutter_monnify.dart';

class MonnifyWebview extends StatefulWidget {
  ///Transaction payload
  final Map<String, dynamic> payload;

  ///display toast to user based on the various state
  final bool displayToast;

  /// Customize how toast is displayed
  final ToastConfig? toast;

  /// Customize look of the appbar
  final AppBarConfig? appBar;

  const MonnifyWebview(this.payload,
      {Key? key, this.displayToast = true, this.toast, this.appBar})
      : super(key: key);
  @override
  State<MonnifyWebview> createState() => _MonnifyWebviewState();
}

class _MonnifyWebviewState extends State<MonnifyWebview>
    with WidgetsBindingObserver {
  late WebViewPlusController _controller;
  double loadProgress = 0;
  String url = "https://monnify-web-asset.vercel.app";

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(
          widget.appBar?.title ?? "Monnify",
          style: TextStyle(color: widget.appBar?.titleColor),
        ),
        elevation: 0,
        backgroundColor: widget.appBar?.backgroundColor,
        leading: widget.appBar?.leadingIcon ??
            IconButton(
              onPressed: () {
                Navigator.pop(context);
              },
              icon: const Icon(
                Icons.chevron_left,
                // color: baseBlack,
              ),
            ),
        actions: [
          IconButton(
            onPressed: () {
              _controller.loadUrl(url);
            },
            icon: const Icon(Icons.refresh
                // color: baseBlack,
                ),
          ),
        ],
      ),
      body: SafeArea(
        child: SizedBox(
            width: deviceWidth(context),
            height: deviceHeight(context),
            child: Column(
              children: [
                loadProgress < 1.0
                    ? Expanded(
                        flex: loadProgress < 1.0 ? 10 : 0,
                        child: _loadingView())
                    : Container(),
                Expanded(flex: 1, child: _mainView()),
              ],
            )),
      ),
    );
  }

  Widget _loadingView() {
    return Container(
        width: 40,
        height: 40,
        padding: const EdgeInsets.symmetric(vertical: 36),
        child: const Center(child: CircularProgressIndicator.adaptive()));
  }

  Widget _mainView() {
    return WebViewPlus(
      initialUrl: url,
      onWebViewCreated: (WebViewPlusController controller) {
        _controller = controller;
        controller.loadUrl(url);
      },
      onPageFinished: (String url) async {
        await _controller.webViewController
            .runJavascript('showpaymentModal(${json.encode(widget.payload)})');
      },
      onProgress: (int progress) {
        setState(() {
          loadProgress = progress / 100;
        });
      },
      debuggingEnabled: true,
      javascriptMode: JavascriptMode.unrestricted,
      javascriptChannels: <JavascriptChannel>{
        _monnifySDKJavascriptChannel(),
      },
    );
  }

  JavascriptChannel _monnifySDKJavascriptChannel() {
    return JavascriptChannel(
        name: 'monnifyFlutterClient',
        onMessageReceived: (JavascriptMessage message) {
          Map<String, dynamic> response = json.decode(message.message);
          if (kDebugMode) {
            print(response["message"]);
            print(response["data"]);
            print(response.toString());
            print("========");
          }
          returnTransactionStatus(response);
        });
  }

  returnTransactionStatus(Map<String, dynamic> response) {
    String toastMessage = response["message"] ?? "";

    //returns to the calling screen if "returnToCaller" is true
    if (response.containsKey("data") &&
        response.containsKey("returnToCaller") &&
        response["returnToCaller"]) {
      late TransactionResponse transactionResponse;

      Map<String, dynamic> responseData = response["data"];

      //transaction was cancelled
      if (responseData.containsKey("paymentStatus")) {
        toastMessage = responseData["paymentStatus"];

        transactionResponse = TransactionResponse(
            status: responseData["paymentStatus"],
            message: responseData["responseMessage"],
            amount: responseData["authorizedAmount"],
            rawData: responseData);
      }
      if (responseData.containsKey("status")) {
        toastMessage = responseData["message"] ?? "";
        if (responseData.containsKey("payload")) {
          transactionResponse = TransactionResponse(
              status: responseData["status"],
              message: responseData["message"] ?? responseData["status"],
              amount: responseData["payload"]["amount"],
              transactionReference: responseData["payload"]
                  ["transactionReference"],
              paymentReference: responseData["payload"]["paymentReference"],
              currency: responseData["payload"]["currency"],
              rawData: responseData);
        } else {
          transactionResponse = TransactionResponse(
              status: responseData["status"],
              message: responseData["message"] ?? responseData["status"],
              amount: responseData["authorizedAmount"],
              transactionReference: responseData["transactionReference"],
              paymentReference: responseData["paymentReference"],
              rawData: responseData);
        }
      }
      //show toast message
      if (widget.displayToast) {
        showToast(toastMessage);
      }

      //return transaction response
      if (Navigator.canPop(context)) {
        Navigator.pop(context, transactionResponse);
      }
    } else {
      if (widget.displayToast) {
        showToast(toastMessage);
      }
    }
  }

  showToast(String message) {
    Fluttertoast.showToast(
        msg: message,
        toastLength: widget.toast?.length ?? Toast.LENGTH_LONG,
        gravity: widget.toast?.position ?? ToastGravity.BOTTOM,
        timeInSecForIosWeb: 1,
        backgroundColor: widget.toast?.backgroundColor ?? Colors.grey,
        textColor: widget.toast?.color ?? Colors.black,
        fontSize: widget.toast?.fontSize ?? 16.0);
  }
}


assets/web/monnify_checkout.js

function postToFlutterClient(data){
    monnifyFlutterClient.postMessage(JSON.stringify(data));
}

function showpaymentModal(monnifyPayload){
    initializeSDK(monnifyPayload)
}

function initializeSDK(payload) {
    try{    
        MonnifySDK.initialize({
            amount: payload.amount, 
            currency: payload.currency,
            reference: payload.reference,
            customerFullName: payload.customerFullName,
            customerEmail: payload.customerEmail,
            apiKey: payload.apiKey,
            contractCode: payload.contractCode,
            paymentDescription: payload.paymentDescription,
            metadata: payload.metadata || null,
            incomeSplitConfig: payload.incomeSplitConfig || [],
            paymentMethods: payload.paymentMethods || [],
            onLoadStart: () => {
                //SDK is initialized
                postToFlutterClient({"returnToCaller": false, "message": "Initializing SDK..."})
            },
            onLoadComplete: () => {
                //SDK is loaded
                postToFlutterClient({"returnToCaller": false, "message": "SDK initialized!!!"})
            },
            onComplete: function(response) {
                //Transaction is completed.
                postToFlutterClient({"returnToCaller": true, "message": "Success", "data": response})
            },
            onClose: function(data) {
                //Modal is closed here
                postToFlutterClient({"returnToCaller": true, "message": "", "data": data})
            }
        });
    }
    catch(e){
        data = {
            "status": "SDK_INITIALIZATION_FAILED",
            "responseCode":"SDK_INITIALIZATION_FAILED",
            "paymentStatus":"SDK_INITIALIZATION_FAILED",
            "responseMessage":`Error initializing SDK. Verify you are sending the right initialization payload: ${e}`,
        }
        postToFlutterClient(data);
    }
}



lib/src/model/transaction_response.dart
import 'dart:convert';

TransactionResponse transactionResponseFromJson(String str) =>
    TransactionResponse.fromJson(json.decode(str));

String transactionResponseToJson(TransactionResponse data) =>
    json.encode(data.toJson());

class TransactionResponse {
  TransactionResponse(
      {this.status,
      this.message,
      this.transactionReference,
      this.paymentReference,
      this.amount,
      this.currency = "NGN",
      this.rawData});

  String? status;
  String? message;
  String? transactionReference;
  String? paymentReference;
  int? amount;
  String? currency;
  dynamic rawData;

  factory TransactionResponse.fromJson(Map<String, dynamic> json) =>
      TransactionResponse(
        status: json["status"],
        message: json["message"],
        transactionReference: json["transactionReference"],
        paymentReference: json["paymentReference"],
        amount: json["authorizedAmount"],
        currency: json["currency"],
      );

  Map<String, dynamic> toJson() => {
        "status": status,
        "message": message,
        "transactionReference": transactionReference,
        "paymentReference": paymentReference,
        "amount": amount,
        "currency": currency
      };
}


example/lib/main.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_monnify/flutter_monnify.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Monnify',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const PayWithMonnify(),
    );
  }
}

class PayWithMonnify extends StatefulWidget {
  const PayWithMonnify({super.key});

  @override
  State<PayWithMonnify> createState() => _PayWithMonnifyState();
}

class _PayWithMonnifyState extends State<PayWithMonnify> {
  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    double height = MediaQuery.of(context).size.height;
    return Scaffold(
        body: SizedBox(
      width: width,
      height: height,
      child: Center(
        child: TextButton(
          child: const Text("Pay With Monnify SDK"),
          onPressed: () async {
            TransactionResponse? response = await Monnify().checkout(
                context, monnifyPayload(),
                appBar: AppBarConfig(
                    titleColor: Colors.white, backgroundColor: Colors.red),
                toast: ToastConfig(
                    color: Colors.black, backgroundColor: Colors.red));

            //call the backend to verify transaction status before providing value
            if (kDebugMode) {
              print("Future completed======>${response?.toJson().toString()}");
              print("Future completed11======>${response?.message.toString()}");
            }
          },
        ),
      ),
    ));
  }

  Map<String, dynamic> monnifyPayload() {
    return {
      "amount": 100,
      "currency": "NGN",
      "reference": DateTime.now().toIso8601String(),
      "customerFullName": "John Doe",
      "customerEmail": "customer@gmail.com",
      "apiKey": "MK_TEST_595UN92CCV",
      "contractCode": "8628159341",
      "paymentDescription": "Lahray World",
      "metadata": {"name": "Damilare", "age": 45},
      // "incomeSplitConfig": [
      //   {
      //     "subAccountCode": "MFY_SUB_342113621921",
      //     "feePercentage": 50,
      //     "splitAmount": 1900,
      //     "feeBearer": true
      //   },
      //   {
      //     "subAccountCode": "MFY_SUB_342113621922",
      //     "feePercentage": 50,
      //     "splitAmount": 2100,
      //     "feeBearer": true
      //   }
      // ],
      "paymentMethod": ["CARD", "ACCOUNT_TRANSFER", "USSD", "PHONE_NUMBER"],
    };
  }
}
